File: problems/树/109.有序链表转换二叉搜索树.cpp
#include <iostream>
#include <functional>
using namespace std;
struct TreeNode {
    int val;
    TreeNode* left;
    TreeNode* right;
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
};

struct ListNode {
    int val;
    ListNode* next;
    ListNode(int x) : val(x), next(nullptr) {}
};
TreeNode* sortedListToBST(ListNode* head) {
    if (!head) return nullptr;
    if (!head->next) return new TreeNode(head->val);
    ListNode *slow = head, *fast = head, *prev = nullptr;
    while (fast && fast->next) {
        prev = slow; slow = slow->next; fast = fast->next->next;
    }
    TreeNode* root = new TreeNode(slow->val);
    if (prev) { prev->next = nullptr; root->left = sortedListToBST(head); }
    root->right = sortedListToBST(slow->next);
    return root;
}
int main() {
    ListNode* head = new ListNode(-10);
    head->next = new ListNode(-3);
    head->next->next = new ListNode(0);
    head->next->next->next = new ListNode(5);
    head->next->next->next->next = new ListNode(9);
    TreeNode* root = sortedListToBST(head);
    cout << root->val << endl;  // 0
    cout << (sortedListToBST(nullptr) == nullptr ? "null" : "?") << endl;  // null
    return 0;
}

File: problems/树/109.有序链表转换二叉搜索树.py
#!/usr/bin/env python3
"""
109. 有序链表转换二叉搜索树
"""
from typing import Optional
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

def sorted_list_to_bst(head: Optional[ListNode]) -> Optional[TreeNode]:
    if not head: return None
    if not head.next: return TreeNode(head.val)
    slow = fast = head
    prev = None
    while fast and fast.next:
        prev = slow
        slow = slow.next
        fast = fast.next.next
    root = TreeNode(slow.val)
    if prev:
        prev.next = None
        root.left = sorted_list_to_bst(head)
    root.right = sorted_list_to_bst(slow.next)
    return root

if __name__ == "__main__":
    # 构建链表: -10 -> -3 -> 0 -> 5 -> 9
    head1 = ListNode(-10, ListNode(-3, ListNode(0, ListNode(5, ListNode(9)))))
    root1 = sorted_list_to_bst(head1)
    print(root1.val)  # 0
    print(sorted_list_to_bst(None))  # None
